src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,462 @@ void Frame::denoise(
+
+                // Search layer index in the file.
+                size_t subimage_index(~0);
+                for (size_t s = 0; s < checkpoint_props.size(); ++s)
+                {
+                    if (get<0>(checkpoint_props[s]) == aov_name)
+                    {
+                        subimage_index = s;
+                        break;
+                    }
+                }
+
+                assert(subimage_index != size_t(~0));
+
+                Image& aov_image = aov->get_image();
+                Tile& aov_tile = aov_image.tile(tile_x, tile_y);
+                reader.choose_subimage(subimage_index);
+                reader.read_tile(tile_x, tile_y, &aov_tile);
+            }
+        }
+    }
+
+    // Load internal AOVs (from external files).
+    for (size_t i = 0, e = internal_aovs().size(); i < e; ++i)
+    {
+        AOV* aov = internal_aovs().get_by_index(i);
+
+        DenoiserAOV* denoiser_aov = dynamic_cast<DenoiserAOV*>(aov);
+
+        // Load denoiser checkpoint.
+        if (denoiser_aov != nullptr)
+        {
+            if (!load_denoiser_checkpoint(impl->m_checkpoint_resume_path, denoiser_aov))
+                return false;
+        }
+    }
+
+    RENDERER_LOG_INFO("read checkpoint file %s, resuming rendering at pass %s.",
+        impl->m_checkpoint_resume_path.c_str(),
+        pretty_uint(start_pass).c_str());
+
+    impl->m_initial_pass = start_pass;
+
+    return true;
+}
+
+void Frame::save_checkpoint(
+    PermanentShadingResultFrameBufferFactory*   buffer_factory,
+    const size_t                                pass) const
+{
+    if (!impl->m_checkpoint_create)
+        return;
+
+    create_parent_directories(impl->m_checkpoint_create_path.c_str());
+
+    GenericImageFileWriter writer(impl->m_checkpoint_create_path.c_str());
+
+    ImageAttributes image_attributes = ImageAttributes::create_default_attributes();
+    image_attributes.insert("appleseed:LastPass", pass);
+
+    // Create layers.
+    // Buffer containing pixels' weight.
+    ShadingBufferCanvas weight_canvas(*this, buffer_factory);
+
+    // Create channel names.
+    const size_t shading_channel_count = weight_canvas.properties().m_channel_count;
+    vector<string> shading_channel_names; // for shading buffer
+    vector<const char *> shading_channel_names_cr;
+
+    static const string channel_name_prefix = "channel_";
+    for (size_t i = 0; i < shading_channel_count; ++i)
+    {
+        shading_channel_names.push_back(channel_name_prefix + pad_left(to_string(i + 1), '0', 4));
+        shading_channel_names_cr.push_back(shading_channel_names[i].c_str());
+    }
+
+    assert(shading_channel_names.size() == shading_channel_count);
+
+    // Add layers in the file.
+    writer.append_image(&image());
+    image_attributes.insert("image_name", "beauty");
+    writer.set_image_attributes(image_attributes);
+
+    writer.append_image(&weight_canvas);
+    image_attributes.insert("image_name", "appleseed:RenderingBuffer");
```

Comment:
Reusing `image_attributes` is confusing and potentially dangerous: you might end up writing attributes you didn't intend to. For instance, right now the main image and all AOV images have a `appleseed:LastPass` attribute, is that intended?

If possible, I would suggest create a new `ImageAttributes` object every time. If necessary, you can wrap each block in braces (`{}`) in order to isolate them from each other.

Reply:
Not it wasn't intended, thanks for pointing it out